File: backend/rpc/handler/data/job_level.py
from rpc.gen.data.job_level.services import TDataJobLevelService
from rpc.gen.data.job_level.structs.ttypes import TJobLevel

from rpc.gen.user.user.types.ttypes import TUserRole

from models import models

class TDataJobLevelServiceHandler(TDataJobLevelService.Iface):
    def __init__(self):
        self.auth_model = models['auth']
        self.job_level_model = models['job_level']

    def fetch(self, auth_token, query=None):
        auth_payload = self.auth_model.decode_auth(auth_token)
        job_levels = self.job_level_model.fetch(query)
        return [x.fill(TJobLevel()) for x in job_levels]

    def get(self, auth_token, job_level_id):
        auth_payload = self.auth_model.decode_auth(auth_token)
        return self.job_level_model.get(job_level_id).fill(TJobLevel())

    def create(self, auth_token, form):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.create(form)
        self.job_level_model.commit()
        return job_level.fill(TJobLevel())

    def set_enabled(self, auth_token, job_level_id, enabled):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_enabled(job_level, enabled)
        self.job_level_model.commit()

    def set_nama(self, auth_token, job_level_id, nama):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_nama(job_level, nama)
        self.job_level_model.commit()

    def set_gaji_pokok(self, auth_token, job_level_id, gaji_pokok):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_gaji_pokok(job_level, gaji_pokok)
        self.job_level_model.commit()

    def set_tunjangan_jabatan(self, auth_token, job_level_id, tunjangan_jabatan):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_tunjangan_jabatan(job_level, tunjangan_jabatan)
        self.job_level_model.commit()

    def set_gaji_pokok(self, auth_token, job_level_id, gaji_pokok):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_gaji_pokok(job_level, gaji_pokok)
        self.job_level_model.commit()

    def set_upah_lembur(self, auth_token, job_level_id, upah_lembur):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_upah_lembur(job_level, upah_lembur)
        self.job_level_model.commit()

    def set_upah_lembur_1(self, auth_token, job_level_id, upah_lembur):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_upah_lembur_1(job_level, upah_lembur)
        self.job_level_model.commit()

    def set_upah_lembur_2(self, auth_token, job_level_id, upah_lembur):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_upah_lembur_2(job_level, upah_lembur)
        self.job_level_model.commit()

    def set_upah_lembur_3(self, auth_token, job_level_id, upah_lembur):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        job_level = self.job_level_model.get(job_level_id)
        self.job_level_model.set_upah_lembur_3(job_level, upah_lembur)
        self.job_level_model.commit()

File: frontend/src/rpc/client/data/job_level.js
import TDataJobLevelService from '@/rpc/gen/TDataJobLevelService';
import { TBaseClient } from '@/rpc/client/base';
import { TUserRole } from '@/rpc/gen/user.user.types_types';


class TDataJobLevelClient extends TBaseClient{
	constructor(stores=null){
		super(stores, TDataJobLevelService, '/api/data/job-level');
	}

	async fetch(query=null){
		return await this.client.fetch(this.stores.auth.authToken, query);
	}
	async get(job_level_id){
		return await this.client.get(this.stores.auth.authToken, job_level_id);
	}

	async create(form){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		return await this.client.create(this.stores.auth.authToken, form);
	}
	async set_name(job_level_id, new_name){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_name(this.stores.auth.authToken, job_level_id, new_name);
	}
	async set_enabled(job_level_id, new_enabled){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_enabled(this.stores.auth.authToken, job_level_id, new_enabled);
	}
	async set_gaji_pokok(job_level_id, new_gaji_pokok){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_gaji_pokok(this.stores.auth.authToken, job_level_id, new_gaji_pokok);
	}
	async set_tunjangan_jabatan(job_level_id, new_tunjangan_jabatan){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_tunjangan_jabatan(this.stores.auth.authToken, job_level_id, new_tunjangan_jabatan);
	}
	async set_upah_lembur(job_level_id, new_upah_lembur, i){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_upah_lembur(this.stores.auth.authToken, job_level_id, new_upah_lembur, i);
	}
	async set_upah_lembur_1(job_level_id, new_upah_lembur){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_upah_lembur_1(this.stores.auth.authToken, job_level_id, new_upah_lembur);
	}
	async set_upah_lembur_2(job_level_id, new_upah_lembur){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_upah_lembur_2(this.stores.auth.authToken, job_level_id, new_upah_lembur);
	}
	async set_upah_lembur_3(job_level_id, new_upah_lembur){
		this.stores.helper.auth.requireRole(TUserRole.SUPER_ADMIN);
		await this.client.set_upah_lembur_3(this.stores.auth.authToken, job_level_id, new_upah_lembur);
	}
}

const dataJobLevelClient = new TDataJobLevelClient();

export { TDataJobLevelClient, dataJobLevelClient }
export default dataJobLevelClient